jni-gen/builtins/utils.rs

```diff
@@ -0,0 +1,39 @@
+use jni::{errors::Result, objects::JObject, JNIEnv};
+use jni::strings::JavaStr;
+use std::ffi::CStr;
+
+fn java_str_to_string(string: &JavaStr) -> Result<String> {
+    unsafe { match CStr::from_ptr(string.get_raw()).to_str() {
+        Ok(resultstring) => Ok(resultstring.to_string()),
+        Err(_) => Ok("Cannot be determined".to_string()),
+    } }
+}
+
+fn get_jobject_class_name_inner<'a, 'c, O>(env: &'c JNIEnv<'c>, object: O) -> Result<String>
+where
+    O: Into<JObject<'a>>
+{
+    let object_class = env.get_object_class(object)?;
+    let res = java_str_to_string(
+        &env.get_string(
+            env.call_method(object_class, "getName", "()Ljava/lang/String;", &[])?
+                 .l()?
+                 .into(),
+        )?,
+    );
+
+    match res {
+        Ok(str) => Ok(str.replace(".", "/")),
+        Err(_) => Ok("Cannot be determined".to_string()),
+    }
+}
+
+pub fn get_jobject_class_name<'a, O>(env: &'a JNIEnv<'a>, object: O) -> String
+where
+    O: Into<JObject<'a>>
+{
+    match get_jobject_class_name_inner(env, object) {
+        Ok(result) => result,
+        Err(_) => "Cannot be determined".to_string(),
+    }
+}
```

Comment:
I had issues with Error types conversions when not having a dependency on `jni-gen`

Reply:
Could you show one of those errors?

Reply:
Maybe you miss something like a `use jni::errors::{Error, Result as JniResult};`

Reply:
My problems, in this case, were that some things are reliant on `jni-gen`

```
pub fn java_str_to_string(string: &JavaStr) -> JniResult<String> {
    unsafe { Ok(CStr::from_ptr(string.get_raw()).to_str()?.to_string()) }
}
```
fails with:
```
error[E0277]: `?` couldn't convert the error to `jni::errors::Error`
  --> jni-gen/systest/src/../gen/builtins/utils.rs:14:58
   |
13 | pub fn java_str_to_string(string: &JavaStr) -> JniResult<String> {
   |                                                ----------------- expected `jni::errors::Error` because of this
14 |     unsafe { Ok(CStr::from_ptr(string.get_raw()).to_str()?.to_string()) }
   |                                                          ^ the trait `From<Utf8Error>` is not implemented for `jni::errors::Error`
```
Which can be solved with `use jni_gen::errors::*`, but this requires then to have `jni_gen` as a regular (not just build) dependency for both systest and vyper-sys. Is that ok? The change would look like this (pushed to this branch) https://github.com/viperproject/prusti-dev/pull/1364/commits/781ea03369e27954742a68826525d5d25c7bb9f9

Reply:
Adding a dependency on `jni-gen` is not necessary and would break the high-level design of `jni-gen`; I would avoid that. Your issue is that `to_str` returns a `Result<&str, Utf8Error>`, but you declared the return type of `java_str_to_string` to be `Result<String, jni::errors::Error>`. The compiler complains that there is no automatic convertsion from `Utf8Error` to `jni::errors::Error`. Since `utils.rs` is only used in debug assertions I think the easier solution is to replace every `?` of non-jni errors with an `.unwrap_or_else(|err| panic!("Failed to ...: {err:?}"))`.

Reply:
Removed the dependency